File: libraries/expo-ondevice-ai/ios/ExpoOndeviceAiHelper.swift
import Locanara

/// Decodes JS options dictionaries into chain constructor parameters
@available(iOS 15.0, macOS 14.0, *)
enum ExpoOndeviceAiHelper {

    // MARK: - Summarize

    static func bulletCount(from options: [String: Any]?) -> Int {
        guard let opts = options else { return 1 }
        if let outputType = opts["outputType"] as? String {
            switch outputType {
            case "TWO_BULLETS": return 2
            case "THREE_BULLETS": return 3
            default: return 1
            }
        }
        return 1
    }

    // MARK: - Classify

    static func classifyOptions(from options: [String: Any]?) -> (categories: [String], maxResults: Int) {
        guard let opts = options else {
            return (["positive", "negative", "neutral"], 3)
        }
        let categories = (opts["categories"] as? [String]) ?? ["positive", "negative", "neutral"]
        let maxResults = (opts["maxResults"] as? Int) ?? 3
        return (categories, maxResults)
    }

    // MARK: - Extract

    static func entityTypes(from options: [String: Any]?) -> [String] {
        (options?["entityTypes"] as? [String]) ?? ["person", "location", "date", "organization"]
    }

    // MARK: - Chat

    static func chatOptions(from options: [String: Any]?) -> (systemPrompt: String, memory: (any Memory)?) {
        let systemPrompt = (options?["systemPrompt"] as? String) ?? "You are a friendly, helpful assistant."

        var memory: (any Memory)? = nil
        if let historyArray = options?["history"] as? [[String: String]], !historyArray.isEmpty {
            memory = PrefilledMemory(history: historyArray)
        }

        return (systemPrompt, memory)
    }

    // MARK: - Translate

    static func translateOptions(from options: [String: Any]?) -> (sourceLanguage: String, targetLanguage: String) {
        let source = (options?["sourceLanguage"] as? String) ?? "en"
        let target = (options?["targetLanguage"] as? String) ?? "en"
        return (source, target)
    }

    // MARK: - Rewrite

    static func rewriteStyle(from options: [String: Any]?) -> RewriteOutputType {
        guard let opts = options,
              let outputTypeStr = opts["outputType"] as? String,
              let outputType = RewriteOutputType(rawValue: outputTypeStr) else {
            return .rephrase
        }
        return outputType
    }
}

// MARK: - Prefilled Memory

/// Memory adapter that provides pre-filled chat history from JS.
@available(iOS 15.0, macOS 14.0, *)
final class PrefilledMemory: Memory, @unchecked Sendable {
    private let entries: [MemoryEntry]

    init(history: [[String: String]]) {
        self.entries = history.compactMap { msg in
            guard let role = msg["role"], let content = msg["content"] else { return nil }
            return MemoryEntry(role: role, content: content)
        }
    }

    func load(for input: ChainInput) async -> [MemoryEntry] { entries }
    func save(input: ChainInput, output: ChainOutput) async { }
    func clear() async { }

    var estimatedTokenCount: Int {
        entries.reduce(0) { $0 + ($1.content.count / 4) }
    }
}

File: libraries/expo-ondevice-ai/android/src/main/java/expo/modules/ondeviceai/ExpoOndeviceAiHelper.kt
package expo.modules.ondeviceai

import com.locanara.ChatMessageInput
import com.locanara.ChatParametersInput
import com.locanara.ProofreadInputType
import com.locanara.RewriteOutputType
import com.locanara.SummarizeInputType
import com.locanara.SummarizeOutputType

/** Decodes JS options maps into chain constructor parameters */
object ExpoOndeviceAiHelper {
    // region Summarize

    fun summarizeOutputType(options: Map<String, Any>?): SummarizeOutputType =
        when (options?.get("outputType") as? String) {
            "TWO_BULLETS" -> SummarizeOutputType.TWO_BULLETS
            "THREE_BULLETS" -> SummarizeOutputType.THREE_BULLETS
            else -> SummarizeOutputType.ONE_BULLET
        }

    fun summarizeInputType(options: Map<String, Any>?): SummarizeInputType =
        when (options?.get("inputType") as? String) {
            "CONVERSATION" -> SummarizeInputType.CONVERSATION
            else -> SummarizeInputType.ARTICLE
        }

    // endregion

    // region Classify

    fun classifyOptions(options: Map<String, Any>?): Pair<List<String>, Int> {
        @Suppress("UNCHECKED_CAST")
        val categories =
            (options?.get("categories") as? List<String>)
                ?: listOf("positive", "negative", "neutral")
        val maxResults = (options?.get("maxResults") as? Number)?.toInt() ?: 3
        return Pair(categories, maxResults)
    }

    // endregion

    // region Extract

    fun extractOptions(options: Map<String, Any>?): Pair<List<String>, Boolean> {
        @Suppress("UNCHECKED_CAST")
        val entityTypes =
            (options?.get("entityTypes") as? List<String>)
                ?: listOf("person", "location", "date", "organization")
        val extractKeyValues = options?.get("extractKeyValues") as? Boolean ?: false
        return Pair(entityTypes, extractKeyValues)
    }

    // endregion

    // region Chat

    @Suppress("UNCHECKED_CAST")
    fun chatParameters(options: Map<String, Any>?): ChatParametersInput {
        val systemPrompt =
            (options?.get("systemPrompt") as? String)
                ?: "You are a friendly, helpful assistant."

        val historyArray = options?.get("history") as? List<Map<String, String>>
        val history =
            historyArray?.mapNotNull { message ->
                val role = message["role"] ?: return@mapNotNull null
                val content = message["content"] ?: return@mapNotNull null
                ChatMessageInput(role = role, content = content)
            }

        return ChatParametersInput(
            conversationId = options?.get("conversationId") as? String,
            systemPrompt = systemPrompt,
            history = history,
        )
    }

    // endregion

    // region Translate

    fun translateOptions(options: Map<String, Any>?): Pair<String, String> {
        val source = (options?.get("sourceLanguage") as? String) ?: "en"
        val target = (options?.get("targetLanguage") as? String) ?: "en"
        return Pair(source, target)
    }

    // endregion

    // region Rewrite

    fun rewriteStyle(options: Map<String, Any>?): RewriteOutputType {
        val outputType = options?.get("outputType") as? String
        return outputType?.let {
            runCatching { RewriteOutputType.valueOf(it) }.getOrNull()
        } ?: RewriteOutputType.REPHRASE
    }

    // endregion

    // region Proofread

    fun proofreadInputType(options: Map<String, Any>?): ProofreadInputType =
        when (options?.get("inputType") as? String) {
            "VOICE" -> ProofreadInputType.VOICE
            else -> ProofreadInputType.KEYBOARD
        }

    // endregion
}
